File: jvm/src/test/kotlin/shareandcharge/openchargingnetwork/notary/TestCases.kt
package shareandcharge.openchargingnetwork.notary

import com.fasterxml.jackson.annotation.JsonProperty

data class SigningTestCase(val input: ValuesToSign<*>, val keys: List<String>, val values: List<String>)

data class ExampleDataClass(@JsonProperty("is_valid") val isValid: Boolean, val message: String)

val mockHeaders = SignableHeaders("456", "DE", "ABC", "DE", "XYZ")

val headersJsonPaths = listOf(
        "x-correlation-id",
        "ocpi-from-country-code",
        "ocpi-from-party-id",
        "ocpi-to-country-code",
        "ocpi-to-party-id").map { "$['headers']['$it']" }

val headersValues = listOf("456", "DE", "ABC", "DE", "XYZ")

val signingTestCases = listOf(
        arrayOf("string values", SigningTestCase(
                input = ValuesToSign(mockHeaders, body = mapOf("id" to "1", "city" to "Essen")),
                keys = headersJsonPaths + listOf("['id']", "['city']").map { "$['body']$it" },
                values = headersValues + listOf("1", "Essen"))),
        arrayOf("mixedValues", SigningTestCase(
                input = ValuesToSign(mockHeaders, body = mapOf("id" to "1", "city" to "Essen", "valid" to true, "count" to 3)),
                keys = headersJsonPaths + listOf("['id']", "['city']", "['valid']", "['count']").map { "$['body']$it" },
                values = headersValues + listOf("1", "Essen", "true", "3"))),
        arrayOf("array values", SigningTestCase(
                input = ValuesToSign(mockHeaders, body = mapOf(
                        "id" to "1",
                        "city" to "Essen",
                        "types" to listOf("WHITELIST", "ALWAYS"))),
                keys = headersJsonPaths + listOf("['id']", "['city']", "['types'][0]", "['types'][1]").map { "$['body']$it" },
                values = headersValues + listOf("1", "Essen", "WHITELIST", "ALWAYS"))),
        arrayOf("objects within arrays", SigningTestCase(
                input = ValuesToSign(mockHeaders, body = mapOf(
                        "id" to "1",
                        "city" to "Essen",
                        "evses" to setOf(mapOf(
                                "id" to "1234",
                                "status" to "BLOCKED")))),
                keys = headersJsonPaths + listOf("['id']", "['city']", "['evses'][0]['id']", "['evses'][0]['status']").map { "$['body']$it" },
                values = headersValues + listOf("1", "Essen", "1234", "BLOCKED"))),
        arrayOf("nested objects within arrays", SigningTestCase(
                input = ValuesToSign(mockHeaders, body = mapOf(
                        "id" to "1",
                        "city" to "Essen",
                        "evses" to arrayOf(mapOf(
                                "id" to "1234",
                                "status" to "BLOCKED",
                                "connectors" to arrayOf(mapOf("id" to "1")))))),
                keys = headersJsonPaths + listOf("['id']", "['city']", "['evses'][0]['id']", "['evses'][0]['status']", "['evses'][0]['connectors'][0]['id']").map { "$['body']$it" },
                values = headersValues + listOf("1", "Essen", "1234", "BLOCKED", "1"))),
        arrayOf("null value", SigningTestCase(
                input = ValuesToSign(mockHeaders, params = null, body = mapOf("id" to "", "city" to "Essen")),
                keys = headersJsonPaths + listOf("$['body']['city']"),
                values = headersValues + listOf("Essen"))),
        arrayOf("custom value", SigningTestCase(
                input = ValuesToSign(mockHeaders, body = ExampleDataClass(false, "i'm a teapot")),
                keys = headersJsonPaths + listOf("$['body']['is_valid']", "$['body']['message']"),
                values = headersValues + listOf("false", "i'm a teapot"))))

val verifyTestCase = ValuesToSign(mockHeaders, body = mapOf(
        "id" to "1",
        "city" to "Essen",
        "evses" to listOf(mapOf(
                "id" to "1234",
                "status" to "BLOCKED",
                "connectors" to listOf(mapOf("id" to "1"))))))

File: js/test/test.cases.ts
import { IValuesToSign } from "../src/ocpi-request.interface"

interface ISigningTestCase {
    name: string,
    input: IValuesToSign,
    keys: string[]
    values: string[]
}

export const mockHeaders = {
    "x-correlation-id": "456",
    "ocpi-from-country-code": "DE",
    "ocpi-from-party-id": "ABC",
    "ocpi-to-country-code": "DE",
    "ocpi-to-party-id": "XYZ"
}

export const signTestCases: ISigningTestCase[] = [
    {
        name: "string values",
        input: {
            headers: mockHeaders,
            body: { id: "1", city: "Essen" }
        },
        keys: ["['id']", "['city']"],
        values: ["1", "Essen"]
    },
    {
        name: "mixed values",
        input: {
            headers: mockHeaders,
            body: { id: 1, city: "Essen", valid: true }
        },
        keys: ["['id']", "['city']", "['valid']"],
        values: ["1", "Essen", "true"]
    },
    {
        name: "nested values",
        input: {
            headers: mockHeaders,
            body: { id: "1", city: "Essen", coordinates: { latitude: 54.002, longitude: -0.783 } }
        },
        keys: ["['id']", "['city']", "['coordinates']['latitude']", "['coordinates']['longitude']"],
        values: ["1", "Essen", "54.002", "-0.783"]
    },
    {
        name: "array values",
        input: {
            headers: mockHeaders,
            body: { id: "1", city: "Essen", types: ["WHITELIST", "ALWAYS"] }
        },
        keys: ["['id']", "['city']", "['types'][0]", "['types'][1]"],
        values: ["1", "Essen", "WHITELIST", "ALWAYS"]
    },
    {
        name: "objects within arrays",
        input: {
            headers: mockHeaders,
            body: { id: "1", city: "Essen", evses: [{ id: "1234", status: "BLOCKED" }] }
        },
        keys: ["['id']", "['city']", "['evses'][0]['id']", "['evses'][0]['status']"],
        values: ["1", "Essen", "1234", "BLOCKED"]
    },
    {
        name: "nested objects within arrays",
        input: {
            headers: mockHeaders,
            body: {
                id: "1", city: "Essen", evses: [{ id: "1234", status: "BLOCKED", connectors: [{ id: "1" }] }]
            }
        },
        keys: ["['id']", "['city']", "['evses'][0]['id']", "['evses'][0]['status']", "['evses'][0]['connectors'][0]['id']"],
        values: ["1", "Essen", "1234", "BLOCKED", "1"]
    },
    {
        name: "null value",
        input: {
            headers: mockHeaders,
            params: undefined,
            body: { id: "", city: "Essen" }
        },
        keys: ["['city']"],
        values: ["Essen"]
    }
]

export const verifyTestCase = {
    input: {
        headers: mockHeaders,
        body: {
            id: "1", city: "Essen", evses: [{ id: "1234", status: "BLOCKED", connectors: [{ id: "1" }] }]
        }
    }
}
